cmd/zc_traverser_local.go

```diff
@@ -56,28 +67,69 @@ func (t *localTraverser) traverse(processor objectProcessor, filters []objectFil
 
 	} else {
 		if t.recursive {
-			err = filepath.Walk(t.fullPath, func(filePath string, fileInfo os.FileInfo, fileError error) error {
-				if fileError != nil {
-					glcm.Info(fmt.Sprintf("Accessing %s failed with error: %s", filePath, fileError))
-					return nil
-				}
-
-				// skip the subdirectories
-				if fileInfo.IsDir() {
-					return nil
-				}
-
-				t.incrementEnumerationCounter()
-
-				// the relative path needs to be computed from the full path
-				computedRelativePath := strings.TrimPrefix(cleanLocalPath(filePath), t.fullPath)
-
-				// leading path separators are trimmed away
-				computedRelativePath = strings.TrimPrefix(computedRelativePath, common.AZCOPY_PATH_SEPARATOR_STRING)
+			// We want to re-queue symlinks up in their evaluated form because filepath.Walk doesn't evaluate them for us.
+			// So, what is the plan of attack?
+			// Because we can't create endless channels, we create an array instead and use it as a queue.
+			// Furthermore, we use a map as a hashset to avoid re-walking any paths we already know.
+			type walkItem struct {
+				fullPath     string // We need the full, symlink-resolved path to walk against
+				relativeBase string // We also need the relative base path we found the symlink at.
+			}
 
-				return processIfPassedFilters(filters, newStoredObject(fileInfo.Name(), computedRelativePath,
-					fileInfo.ModTime(), fileInfo.Size(), nil, blobTypeNA), processor)
-			})
+			var walkQueue = []walkItem{{fullPath: t.fullPath, relativeBase: ""}}
+			var seenPaths = make(map[string]bool)
+
+			for len(walkQueue) > 0 {
+				queueItem := walkQueue[0]
+				walkQueue = walkQueue[1:] // Handle queue as a sliding window over an array
+				err = filepath.Walk(queueItem.fullPath, func(filePath string, fileInfo os.FileInfo, fileError error) error {
+					if fileError != nil {
+						glcm.Info(fmt.Sprintf("Accessing %s failed with error: %s", filePath, fileError))
+						return nil
+					}
+
+					computedRelativePath := strings.TrimPrefix(cleanLocalPath(filePath), cleanLocalPath(queueItem.fullPath))
+					computedRelativePath = cleanLocalPath(filepath.Join(queueItem.relativeBase, computedRelativePath))
+					computedRelativePath = strings.TrimPrefix(computedRelativePath, common.AZCOPY_PATH_SEPARATOR_STRING)
+
+					if fileInfo.Mode()&os.ModeSymlink != 0 {
```

Comment:
nice, consider adding integration tests for symlink support

Reply:
I don't know if smoke tests actually have good enough coverage

Reply:
Will do